File: src/android/app/src/main/java/info/cemu/Cemu/inputoverlay/RectangleButton.java
package info.cemu.Cemu.inputoverlay;

import android.content.res.Resources;

import androidx.annotation.DrawableRes;

public class RectangleButton extends Button {
    int left;
    int top;
    int right;
    int bottom;

    public RectangleButton(Resources resources, @DrawableRes int pressedButtonId, @DrawableRes int notPressedButtonId, ButtonStateChangeListener buttonStateChangeListener, InputOverlaySurfaceView.OverlayButton overlayButton, InputOverlaySettingsProvider.InputOverlaySettings settings) {
        super(resources, pressedButtonId, notPressedButtonId, buttonStateChangeListener, overlayButton, settings);
        configure();
    }

    @Override
    protected void configure() {
        var rect = settings.getRect();
        this.left = rect.left;
        this.top = rect.top;
        this.right = rect.right;
        this.bottom = rect.bottom;
        iconPressed.setBounds(left, top, right, bottom);
        iconNotPressed.setBounds(left, top, right, bottom);
    }

    @Override
    public boolean isInside(int x, int y) {
        return x >= left && x < right && y >= top && y < bottom;
    }

}

File: src/android/app/src/main/java/info/cemu/cemu/emulation/inputoverlay/inputs/RectangleButton.kt
package info.cemu.cemu.emulation.inputoverlay.inputs

import android.graphics.Canvas
import android.graphics.Rect
import info.cemu.cemu.emulation.inputoverlay.fillRoundRectangleWithStroke
import info.cemu.cemu.emulation.inputoverlay.inputs.innerdrawing.ButtonInnerDrawing

class RectangleButton(
    private val innerDrawing: ButtonInnerDrawing,
    onButtonStateChange: (state: Boolean) -> Unit,
    private val alpha: Int,
    rect: Rect,
) : Button(onButtonStateChange, rect) {
    var left: Float = 0f
    var top: Float = 0f
    var right: Float = 0f
    var bottom: Float = 0f

    init {
        configure()
    }

    override fun configure() {
        this.left = rect.left.toFloat()
        this.top = rect.top.toFloat()
        this.right = rect.right.toFloat()
        this.bottom = rect.bottom.toFloat()
        configureColors(alpha)
        innerDrawing.configure(rect, alpha)
    }

    override fun drawInput(canvas: Canvas) {
        val fillColor: Int
        val strokeColor: Int
        if (state) {
            fillColor = activeFillColor
            strokeColor = activeStrokeColor
        } else {
            fillColor = inactiveFillColor
            strokeColor = inactiveStrokeColor
        }
        canvas.fillRoundRectangleWithStroke(
            left,
            top,
            right,
            bottom,
            RECTANGLE_RADIUS,
            paint,
            fillColor,
            strokeColor
        )
        innerDrawing.draw(canvas, state)
    }

    override fun isInside(x: Float, y: Float): Boolean {
        return x in left..<right && y in top..<bottom
    }

    companion object {
        const val RECTANGLE_RADIUS = 5f
    }
}
